### PR TITLE
remote inference connector and virtual model APIs

### PR DESCRIPTION
**Pull request by Zhangxunmt**
### Description
Remote Inference Connector New APIs, including CRUD.
Examples:
~~~
GET /_plugins/_ml/connectors/277XeIgBsNBQ1UoSal6t
DELETE /_plugins/_ml/connectors/277XeIgBsNBQ1UoSal6t
GET /_plugins/_ml/connectors/_search
{
  "query": {
    "match_all": {}
  },
  "size": 1000
}

POST /_plugins/_ml/connectors/_create
{ connector Template }
~~~
Virtual Model deploy and predict APIs to integrate with the detached Connectors.
 
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: common/src/main/java/org/opensearch/ml/common/CommonValue.java [hunk 1]
```diff
@@ -172,4 +180,55 @@ public class CommonValue {
...
+            + DetachedConnector.CONNECTOR_NAME_FIELD
+            + "\" : {\"type\":\"text\",\"fields\":{\"keyword\":{\"type\":\"keyword\",\"ignore_above\":256}}},\n"
+            + "      \""
+            + DetachedConnector.CONNECTOR_VERSION_FIELD
+            + "\" : {\"type\": \"keyword\"},\n"
+            + "      \""
+            + DetachedConnector.CONNECTOR_DESCRIPTION_FIELD
+            + "\" : {\"type\": \"text\"},\n"
+            + "      \""
+            + DetachedConnector.CONNECTOR_PROTOCOL_FIELD
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/CommonValue.java:199]: I think protocol should be keyword, like `http` right?

File: common/src/main/java/org/opensearch/ml/common/connector/template/APISchema.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+import java.security.PrivilegedExceptionAction;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.opensearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
+import static org.opensearch.ml.common.utils.StringUtils.gson;
+
+@Data
+public class APISchema implements ToXContentObject, NamedWriteable {
+    public static final String PARSE_FIELD_NAME = "API_Schema";
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/connector/template/APISchema.java:30]: Let's follow the same naming pattern, change to lower case and connect with "_", 
```suggestion
    public static final String PARSE_FIELD_NAME = "api_schema";
```

File: common/src/main/java/org/opensearch/ml/common/CommonValue.java (hunk 1 above)
Comment (Zhangxunmt) [reply common/src/main/java/org/opensearch/ml/common/CommonValue.java:199]: Yes, updated already.

File: common/src/main/java/org/opensearch/ml/common/connector/template/APISchema.java (hunk 2 above)
Comment (Zhangxunmt) [reply common/src/main/java/org/opensearch/ml/common/connector/template/APISchema.java:30]: Yes will do in the refresh.

File: common/src/main/java/org/opensearch/ml/common/connector/template/ConnectorState.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.ml.common.connector.template;
+
+public enum ConnectorState {
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/connector/template/ConnectorState.java:8]: Why do we need to track the connector state as we already have model state ?

Comment (Zhangxunmt) [reply common/src/main/java/org/opensearch/ml/common/connector/template/ConnectorState.java:8]: Maybe we can reserve this connector state for future "metadata" or "healthcheck" APIs? Once the connector is not healthy, we can mark it as unhealthy? Let's keep it for a while. If we figure that it's not useful at all, we can always delete it.

**Pull request merged**